Google.Api.Gax.Grpc/Rest/RestChannel.cs

```diff
@@ -105,51 +117,97 @@ private async Task AddAuthHeadersAsync(HttpRequestMessage request, RestMethod me
+        /// and the <see cref="Status"/> of the call are two different functions.
+        /// The function to obtain the response is async, but the function to obtain the
+        /// <see cref="Status"/> is not.
+        /// For being able to surface error details in <see cref="Status"/> we need to be
+        /// able to call <see cref="HttpContent.ReadAsStringAsync"/> which is an async method,
+        /// and thus cannot be done, without blocking, on the sync function that obtains the 
+        /// <see cref="Status"/> in the <see cref="AsyncUnaryCall{TResponse}"/>.
+        /// So we need to make async content reading part of sending the call and not part of
+        /// building the TResponse as it is now.
+        /// This class is just a convenient wrapper for passing together the <see cref="HttpResponseMessage"/>
+        /// and it's read response.
+        /// </summary>
+        internal class ReadHttpResponseMessage
+        {
+            private HttpResponseMessage OriginalResponseMessage { get; }
+
+            private readonly string _content;
+            private readonly ExceptionDispatchInfo _readException;
+
+            internal string Content
+            {
+                get
+                {
+                    _readException?.Throw();
+                    return _content;
+                }
+            }
+
+            internal ReadHttpResponseMessage(HttpResponseMessage response, string content) =>
+                (OriginalResponseMessage, _content) = (response, content);
+
+            internal ReadHttpResponseMessage(HttpResponseMessage response, ExceptionDispatchInfo readException) =>
+                (OriginalResponseMessage, _readException) = (response, readException);
+
+            internal Metadata GetHeaders()
+            {
+                // TODO: This could be very wrong. I don't know what headers we should really return, and I don't know about semi-colon joining.
+                var metadata = new Metadata();
+                foreach (var header in OriginalResponseMessage.Headers)
+                {
+                    metadata.Add(header.Key, string.Join(";", header.Value));
+                }
+                return metadata;
+            }
+
+            internal Status GetStatus()
+            {
+                var grpcStatus = OriginalResponseMessage.StatusCode switch
+                {
+                    HttpStatusCode.OK => StatusCode.OK,
+                    HttpStatusCode.Unauthorized => StatusCode.Unauthenticated,
+                    HttpStatusCode.Forbidden => StatusCode.PermissionDenied,
+                    HttpStatusCode.BadRequest => StatusCode.InvalidArgument,
+                    HttpStatusCode.InternalServerError => StatusCode.Internal,
+                    HttpStatusCode.NotFound => StatusCode.NotFound,
+                    HttpStatusCode.Conflict => StatusCode.FailedPrecondition,
+                    // TODO: Others!
```

Comment:
Another thing to sort out before GA.

Reply:
Agreed.
In particular I think we should map `StatusCode.OK` to `HttpResponseMessage.IsSuccessStatusCode`, since this is the measure of success that we use in Apiary.

Reply:
I don't quite follow... but I think the "OK" path is fine at the moment.